<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Potts model · PottsEvolver.jl</title><meta name="title" content="Potts model · PottsEvolver.jl"/><meta property="og:title" content="Potts model · PottsEvolver.jl"/><meta property="twitter:title" content="Potts model · PottsEvolver.jl"/><meta name="description" content="Documentation for PottsEvolver.jl."/><meta property="og:description" content="Documentation for PottsEvolver.jl."/><meta property="twitter:description" content="Documentation for PottsEvolver.jl."/><meta property="og:url" content="https://pierrebarrat.github.io/PottsEvolver.jl/pottsgraph/"/><meta property="twitter:url" content="https://pierrebarrat.github.io/PottsEvolver.jl/pottsgraph/"/><link rel="canonical" href="https://pierrebarrat.github.io/PottsEvolver.jl/pottsgraph/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">PottsEvolver.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Index</a></li><li class="is-active"><a class="tocitem" href>Potts model</a><ul class="internal"><li><a class="tocitem" href="#Structure-and-Fields"><span>Structure and Fields</span></a></li><li><a class="tocitem" href="#Reading-and-writing-PottsGraph"><span>Reading and writing <code>PottsGraph</code></span></a></li><li><a class="tocitem" href="#Basic-Functions"><span>Basic Functions</span></a></li><li><a class="tocitem" href="#Additional-Utilities"><span>Additional Utilities</span></a></li></ul></li><li><a class="tocitem" href="../sampling/">Sampling</a></li><li><a class="tocitem" href="../genetic_code/">Genetic code</a></li><li><a class="tocitem" href="../reference/">Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Potts model</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Potts model</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/PierreBarrat/PottsEvolver.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/master/docs/src/pottsgraph.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="PottsGraph"><a class="docs-heading-anchor" href="#PottsGraph">PottsGraph</a><a id="PottsGraph-1"></a><a class="docs-heading-anchor-permalink" href="#PottsGraph" title="Permalink"></a></h1><p>The <code>PottsGraph</code> structure represents a Potts model, a statistical model used to describe interactions in a system with multiple states (e.g., amino acids in a protein sequence). It is defined by the following fields:</p><ul><li><code>J</code>: A 4-dimensional array of dimensions <code>q x q x L x L</code> representing the coupling parameters between states at different positions.</li><li><code>h</code>: A 2-dimensional array of dimensions <code>q x L</code> representing the local field parameters for each state at each position.</li><li><code>β</code>: The inverse temperature parameter, used during sampling. </li><li><code>alphabet</code>: An optional <code>Alphabet</code> object to map states to characters (e.g., amino acids), see the <a href="https://github.com/PierreBarrat/BioSequenceMappings.jl">BioSequenceMappings.jl</a> package.</li></ul><p><code>q</code> represents the number of states (e.g., 21 for amino acids + gap), and <code>L</code> is the length of the sequence.</p><h2 id="Structure-and-Fields"><a class="docs-heading-anchor" href="#Structure-and-Fields">Structure and Fields</a><a id="Structure-and-Fields-1"></a><a class="docs-heading-anchor-permalink" href="#Structure-and-Fields" title="Permalink"></a></h2><p>The <code>PottsGraph</code> is defined as follows:</p><pre><code class="language-julia hljs">@kwdef mutable struct PottsGraph{T&lt;:AbstractFloat}
    J::Array{T,4}
    h::Array{T,2}
    β::T = 1.0
    alphabet::Union{Nothing, BioAequenceMappings.Alphabet{Char,&lt;:Integer}} = aa_alphabet
end</code></pre><h3 id="Creating-a-PottsGraph"><a class="docs-heading-anchor" href="#Creating-a-PottsGraph">Creating a PottsGraph</a><a id="Creating-a-PottsGraph-1"></a><a class="docs-heading-anchor-permalink" href="#Creating-a-PottsGraph" title="Permalink"></a></h3><p>You can create a <code>PottsGraph</code> with null or random initialization:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; using PottsEvolver</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; L, q = (10, 21);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; g_null = PottsGraph(L, q, Float64; init=:null); # null initialization</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; g_null.J[:, :, 1, 2]</code><code class="nohighlight hljs ansi" style="display:block;">21×21 Matrix{Float64}:
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 ⋮                        ⋮              ⋱       ⋮                        ⋮
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; g_rand = PottsGraph(L, q, Float64; init=:rand); # random initialization</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; size(g_rand)</code><code class="nohighlight hljs ansi" style="display:block;">(L = 10, q = 21)</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; g_rand.J[:, :, 1, 2] # coupling matrix between sites 1 and 2</code><code class="nohighlight hljs ansi" style="display:block;">21×21 Matrix{Float64}:
 -0.018965   -0.122869     0.114586    …  -0.0833305    0.0149923
 -0.0425263   0.127514     0.101529       -0.0285106    0.00147261
 -0.037448   -0.189481     0.162955       -0.0847158   -0.275771
 -0.206062    0.00463166   0.0108194       0.134124    -0.0140208
  0.152885    0.0192616   -0.00239141      0.0614174    0.0934813
 -0.127539   -0.0925439   -0.0595616   …   0.117198    -0.0156659
  0.0458097  -0.0869267    0.00689393     -0.124348    -0.177321
  0.143304    0.0889836   -0.0176166       0.116187     0.23397
 -0.0180081  -0.066315    -0.0597444       0.139423     0.00768104
 -0.0578209  -0.130038     0.0121156       0.246145    -0.108437
  ⋮                                    ⋱                ⋮
 -0.0528967   0.0440188    0.0574943       0.142087    -0.0826536
 -0.0673243  -0.0891129    0.0326741       0.13679     -0.0165905
  0.0604809  -0.243688    -0.033988       -0.0661026    0.13272
 -0.0257537  -0.161975     0.0433778   …  -0.105733    -0.145078
 -0.0690402  -0.10473      0.0146236       0.0988476    0.0862977
 -0.0515727   0.0694687   -0.00475758     -0.0215803    0.134636
  0.0610611  -0.0059545    0.180214       -0.00194814   0.0233343
 -0.0600621   0.200571    -0.0190536      -0.149355     0.0615126
 -0.0344998   0.0774663   -0.0926385   …   0.200345    -0.118362</code></pre><p>It is sometimes easier to view the couplings <code>J</code> as a square matrix. </p><pre><code class="language-julia-repl hljs">julia&gt; J_matrix = PottsEvolver.couplings_as_matrix(g_rand.J);

julia&gt; size(J_matrix) # (L*q x L*q)
(420, 420)

julia&gt; J_matrix ≈ J_matrix&#39; # The coupling matrix should be symetric
true

julia&gt; J_tensor = PottsEvolver.matrix_as_couplings(J_matrix, L, q); # back to tensor

julia&gt; J_tensor ≈ g_rand.J
true</code></pre><div class="admonition is-info" id="Note-e2b6cceb06f709ea"><header class="admonition-header">Note<a class="admonition-anchor" href="#Note-e2b6cceb06f709ea" title="Permalink"></a></header><div class="admonition-body"><p>There is no check that the coupling matrix is symetric, although it should normally be the case for Potts models. </p></div></div><h2 id="Reading-and-writing-PottsGraph"><a class="docs-heading-anchor" href="#Reading-and-writing-PottsGraph">Reading and writing <code>PottsGraph</code></a><a id="Reading-and-writing-PottsGraph-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-and-writing-PottsGraph" title="Permalink"></a></h2><h3 id="Reading-from-a-File"><a class="docs-heading-anchor" href="#Reading-from-a-File">Reading from a File</a><a id="Reading-from-a-File-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-from-a-File" title="Permalink"></a></h3><p>The <code>read_graph</code> function reads a <code>PottsGraph</code> from a file. The file should follow the format </p><pre><code class="nohighlight hljs">J i j a b value # coupling lines
...
h i a value # field lines</code></pre><p>where <code>i</code> and <code>j</code> are sequence positions and <code>a</code>, <code>b</code> are integer states (<em>e.g.</em> representing amino acids). </p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; # Read the graph in the example folder
       g = read_graph(&quot;../../example/toy_potts.dat)</code><code class="nohighlight hljs ansi" style="display:block;">ERROR: ParseError:
<span class="sgr90"># Error @ </span>8;;file:///home/runner/work/PottsEvolver.jl/PottsEvolver.jl/docs/REPL#2:46<span class="sgr90">REPL:2:46</span>8;;
# Read the graph in the example folder
g = read_graph(&quot;../../example/toy_potts.dat)
<span class="sgr90">#                                           </span>
<span class="sgr90">└ ── </span><span class="sgr91">unterminated string literal</span></code></pre><h3 id="Writing-to-a-File"><a class="docs-heading-anchor" href="#Writing-to-a-File">Writing to a File</a><a id="Writing-to-a-File-1"></a><a class="docs-heading-anchor-permalink" href="#Writing-to-a-File" title="Permalink"></a></h3><p>The <code>write</code> function writes the parameters of a <code>PottsGraph</code> to a file.</p><pre><code class="language-julia hljs">write(&quot;path/to/output.txt&quot;, g; sigdigits=5, index_style=1)</code></pre><h2 id="Basic-Functions"><a class="docs-heading-anchor" href="#Basic-Functions">Basic Functions</a><a id="Basic-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-Functions" title="Permalink"></a></h2><h3 id="Energy-of-a-Sequence"><a class="docs-heading-anchor" href="#Energy-of-a-Sequence">Energy of a Sequence</a><a id="Energy-of-a-Sequence-1"></a><a class="docs-heading-anchor-permalink" href="#Energy-of-a-Sequence" title="Permalink"></a></h3><p>The <code>energy</code> function calculates the statistical energy of a sequence. </p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; seq = AASequence(map(aa_alphabet, collect(&quot;ACDEFGHIKLMNPQRSTVWY&quot;)))</code><code class="nohighlight hljs ansi" style="display:block;">AASequence{Int64}([2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21])</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; E = energy(seq, g_rand)</code><code class="nohighlight hljs ansi" style="display:block;">ERROR: ArgumentError: Lengths of sequence and model do not match
length(s) == (size(g)).L must hold. Got
length(s) =&gt; 20
(size(g)).L =&gt; 10</code></pre><h3 id="Gauge-Change"><a class="docs-heading-anchor" href="#Gauge-Change">Gauge Change</a><a id="Gauge-Change-1"></a><a class="docs-heading-anchor-permalink" href="#Gauge-Change" title="Permalink"></a></h3><p>The <code>set_gauge!</code> function allows you to change the gauge of the <code>PottsGraph</code>. Supported gauges include <code>:zero_sum</code> and <code>:lattice_gas</code>.</p><pre><code class="language-julia-repl hljs">julia&gt; PottsEvolver.set_gauge!(g_rand, :zero_sum);

julia&gt; all(≈(0; atol=1e-14), sum(g_rand.h, dims=1)) # sum of h values along each sequence position is zero
true

julia&gt; all(≈(0; atol=1e-14), sum(g_rand.J, dims=[1,2])) #same goes for J
true</code></pre><h2 id="Additional-Utilities"><a class="docs-heading-anchor" href="#Additional-Utilities">Additional Utilities</a><a id="Additional-Utilities-1"></a><a class="docs-heading-anchor-permalink" href="#Additional-Utilities" title="Permalink"></a></h2><h3 id="Profile-Model"><a class="docs-heading-anchor" href="#Profile-Model">Profile Model</a><a id="Profile-Model-1"></a><a class="docs-heading-anchor-permalink" href="#Profile-Model" title="Permalink"></a></h3><p>The <code>profile_model</code> function creates a <code>PottsGraph</code> with only fields that fit the single-site frequencies in a given matrix.</p><pre><code class="language-julia-repl hljs">julia&gt; f1 = rand(21, 10);  # Example frequency matrix - does not have to be normalized

julia&gt; profile = PottsEvolver.profile_model(f1; pc=1e-2);

julia&gt; all(≈(0; atol=1e-14), profile.J)
true</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Index</a><a class="docs-footer-nextpage" href="../sampling/">Sampling »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Friday 9 January 2026 17:57">Friday 9 January 2026</span>. Using Julia version 1.12.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
