<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Reference · PottsEvolver.jl</title><meta name="title" content="Reference · PottsEvolver.jl"/><meta property="og:title" content="Reference · PottsEvolver.jl"/><meta property="twitter:title" content="Reference · PottsEvolver.jl"/><meta name="description" content="Documentation for PottsEvolver.jl."/><meta property="og:description" content="Documentation for PottsEvolver.jl."/><meta property="twitter:description" content="Documentation for PottsEvolver.jl."/><meta property="og:url" content="https://pierrebarrat.github.io/PottsEvolver.jl/reference/"/><meta property="twitter:url" content="https://pierrebarrat.github.io/PottsEvolver.jl/reference/"/><link rel="canonical" href="https://pierrebarrat.github.io/PottsEvolver.jl/reference/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">PottsEvolver.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Index</a></li><li><a class="tocitem" href="../pottsgraph/">Potts model</a></li><li><a class="tocitem" href="../sampling/">Sampling</a></li><li><a class="tocitem" href="../genetic_code/">Genetic code</a></li><li class="is-active"><a class="tocitem" href>Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Reference</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/PierreBarrat/PottsEvolver.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/master/docs/src/reference.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><p>Documentation for <a href="https://github.com/PierreBarrat/PottsEvolver.jl">PottsEvolver</a>.</p><ul><li><a href="#PottsEvolver.aa_alphabet"><code>PottsEvolver.aa_alphabet</code></a></li><li><a href="#PottsEvolver.codon_alphabet"><code>PottsEvolver.codon_alphabet</code></a></li><li><a href="#PottsEvolver.nt_alphabet"><code>PottsEvolver.nt_alphabet</code></a></li><li><a href="#BioSequenceMappings.Alignment-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:AbstractSequence"><code>BioSequenceMappings.Alignment</code></a></li><li><a href="#PottsEvolver.AASequence"><code>PottsEvolver.AASequence</code></a></li><li><a href="#PottsEvolver.AASequence-Tuple{Integer}"><code>PottsEvolver.AASequence</code></a></li><li><a href="#PottsEvolver.BranchLengthMeaning"><code>PottsEvolver.BranchLengthMeaning</code></a></li><li><a href="#PottsEvolver.CodonSequence-Tuple{Int64}"><code>PottsEvolver.CodonSequence</code></a></li><li><a href="#PottsEvolver.CodonSequence-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:Integer"><code>PottsEvolver.CodonSequence</code></a></li><li><a href="#PottsEvolver.NumSequence"><code>PottsEvolver.NumSequence</code></a></li><li><a href="#PottsEvolver.PottsGraph"><code>PottsEvolver.PottsGraph</code></a></li><li><a href="#PottsEvolver.PottsGraph"><code>PottsEvolver.PottsGraph</code></a></li><li><a href="#PottsEvolver.SamplingParameters"><code>PottsEvolver.SamplingParameters</code></a></li><li><a href="#PottsEvolver.energy-Tuple{AbstractVector{&lt;:Integer}, PottsGraph}"><code>PottsEvolver.energy</code></a></li><li><a href="#PottsEvolver.genetic_code-Tuple{PottsEvolver.Codon}"><code>PottsEvolver.genetic_code</code></a></li><li><a href="#PottsEvolver.genetic_code-Tuple{T} where T&lt;:Integer"><code>PottsEvolver.genetic_code</code></a></li><li><a href="#PottsEvolver.mcmc_sample"><code>PottsEvolver.mcmc_sample</code></a></li><li><a href="#PottsEvolver.mcmc_sample-Tuple{Any, TreeTools.Tree, Any}"><code>PottsEvolver.mcmc_sample</code></a></li><li><a href="#PottsEvolver.mcmc_steps!"><code>PottsEvolver.mcmc_steps!</code></a></li><li><a href="#PottsEvolver.read_graph"><code>PottsEvolver.read_graph</code></a></li><li><a href="#PottsEvolver.read_potts_graph"><code>PottsEvolver.read_potts_graph</code></a></li></ul><article><details class="docstring" open="true"><summary id="PottsEvolver.mcmc_steps!"><a class="docstring-binding" href="#PottsEvolver.mcmc_steps!"><code>PottsEvolver.mcmc_steps!</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">mcmc_steps!(
    s::AbstractSequence, g, num_steps::Integer, p::SamplingParameters;
    gibbs_holder, kwargs...
)</code></pre><p>Perform <code>num_steps</code> discrete MCMC steps starting from sequence <code>s</code> and using graph <code>g</code>. The step type (<code>:gibbs</code>, <code>:metropolis</code>) and the interpretation of <code>num_steps</code>     (<code>:changed</code>, <code>:accepted</code>, <code>:proposed</code>) is set using <code>p</code> (see <code>?SamplingParameters</code>). Modifies the input sequence <code>s</code> and returns it.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sampling_core.jl#L5-L15">source</a></section><section><div><pre><code class="language-julia hljs">mcmc_steps!(
    sequence::AbstractSequence, g, Tmax::AbstractFloat, parameters::SamplingParameters;
    kwargs...
)
mcmc_steps!(
    state::CTMCState, g, Tmax::AbstractFloat, step_type::Symbol; kwargs...
)</code></pre><p>Sample <code>g</code> during time <code>Tmax</code> (continuous) from sequence <code>seq</code>. The step type is <code>p.step_type</code> (see <code>?SamplingParameters</code>). Modifies the input sequence <code>s</code> and returns it. Allocates a <code>CTMCState</code> (three matrices of order <code>q*L</code>).</p><p><strong>Notes</strong></p><ul><li>The form with <code>state::CTMCState</code> will use a pre-allocated <code>CTMCState</code> for calculations.</li><li>Expects <code>parameters.sampling_type</code> to be <code>:continuous</code>. Fails if otherwise.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sampling_continuous_core.jl#L75-L92">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.aa_alphabet"><a class="docstring-binding" href="#PottsEvolver.aa_alphabet"><code>PottsEvolver.aa_alphabet</code></a> — <span class="docstring-category">Constant</span></summary><section><div><pre><code class="language-julia hljs">aa_alphabet
nt_alphabet
codon_alphabet</code></pre><p>Default alphabets for <code>PottsEvolver</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/codons.jl#L7-L13">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.codon_alphabet"><a class="docstring-binding" href="#PottsEvolver.codon_alphabet"><code>PottsEvolver.codon_alphabet</code></a> — <span class="docstring-category">Constant</span></summary><section><div><pre><code class="language-julia hljs">aa_alphabet
nt_alphabet
codon_alphabet</code></pre><p>Default alphabets for <code>PottsEvolver</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/codons.jl#L46-L52">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.nt_alphabet"><a class="docstring-binding" href="#PottsEvolver.nt_alphabet"><code>PottsEvolver.nt_alphabet</code></a> — <span class="docstring-category">Constant</span></summary><section><div><pre><code class="language-julia hljs">aa_alphabet
nt_alphabet
codon_alphabet</code></pre><p>Default alphabets for <code>PottsEvolver</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/codons.jl#L16-L22">source</a></section></details></article><article><details class="docstring" open="true"><summary id="BioSequenceMappings.Alignment-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:AbstractSequence"><a class="docstring-binding" href="#BioSequenceMappings.Alignment-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:AbstractSequence"><code>BioSequenceMappings.Alignment</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">Alignment(sequences; alphabet, names, as_codons=true)</code></pre><p>Construct a <code>BioSequenceMappings.Alignment</code> from a set of sequences. If the sequences are <code>AASequence</code>, <code>alphabet</code> defaults to <code>aa_alphabet</code>. If they are <code>CodonSequence</code>, <code>as_codons</code> can be used to decide whether the alignment should store codons or amino acids. <code>alphabet</code> can be determined automatically from this.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sequences.jl#L230-L238">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.AASequence"><a class="docstring-binding" href="#PottsEvolver.AASequence"><code>PottsEvolver.AASequence</code></a> — <span class="docstring-category">Type</span></summary><section><div><pre><code class="language-julia hljs">mutable struct AASequence{T&lt;:Integer} &lt;: AbstractSequence</code></pre><p>Field: <code>seq::Vector{T}</code>. Wrapper around a vector of integers, with implied alphabet <code>PottsEvolver.aa_alphabet</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sequences.jl#L35-L40">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.AASequence-Tuple{Integer}"><a class="docstring-binding" href="#PottsEvolver.AASequence-Tuple{Integer}"><code>PottsEvolver.AASequence</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">AASequence(L; T)</code></pre><p>Return a random <code>AASequence{T}</code> of length <code>L</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sequences.jl#L58-L62">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.BranchLengthMeaning"><a class="docstring-binding" href="#PottsEvolver.BranchLengthMeaning"><code>PottsEvolver.BranchLengthMeaning</code></a> — <span class="docstring-category">Type</span></summary><section><div><pre><code class="language-julia hljs">BranchLengthMeaning

    type::Symbol
    length::Symbol</code></pre><p>Define how the branch lengths should be interpreted when sampling on a tree with a discrete-time algorithm. This is done in two consecutive steps.</p><ol><li>Decide whether the branch length represents mcmc steps or sweeps.<ul><li>If <code>type==:sweep</code>, multiply the branch length by the length of the sequence.</li><li>Else, if <code>type==:step</code>, leave it as is.</li></ul></li><li>Convert the branch length to an integer value which will be the number of mcmc steps.<ul><li>if <code>length==:exact</code>, the branch length is expected to be an integer (<code>rtol=1e-6</code>).</li><li>if <code>length==:round</code>, round it to the closest <code>Int</code>.</li><li>if <code>length==:poisson</code>, sample a poisson distribution with the branch length as mean.</li></ul></li></ol></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sampling_parameters.jl#L10-L27">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.CodonSequence-Tuple{Int64}"><a class="docstring-binding" href="#PottsEvolver.CodonSequence-Tuple{Int64}"><code>PottsEvolver.CodonSequence</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">CodonSequence(L::Int; source=:aa, T)</code></pre><p>Sample <code>L</code> states at random of the type of <code>source</code> (<code>:aa</code> or <code>:codon</code>):</p><ul><li>if <code>:codon</code>, sample codons at random</li><li>if <code>:aa</code>, sample amino acids at random and reverse translate them randomly to matching codons</li></ul><p>Underlying integer type is <code>T</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sequences.jl#L117-L125">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.CodonSequence-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:Integer"><a class="docstring-binding" href="#PottsEvolver.CodonSequence-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:Integer"><code>PottsEvolver.CodonSequence</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">CodonSequence(seq::Vector{Integer}; source=:aa)</code></pre><p>Build a <code>CodonSequence</code> from <code>seq</code>:</p><ul><li>if <code>source==:codon</code>, <code>seq</code> is interpreted as representing codons (see <code>codon_alphabet</code>);</li><li>if <code>source==:aa</code>, <code>seq</code> is interpreted as representing amino acids (see <code>aa_alphabet</code>); matching codons are randomly chosen using the <code>PottsEvolver.reverse_code_rand</code> method.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sequences.jl#L94-L101">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.NumSequence"><a class="docstring-binding" href="#PottsEvolver.NumSequence"><code>PottsEvolver.NumSequence</code></a> — <span class="docstring-category">Type</span></summary><section><div><pre><code class="language-julia hljs">NumSequence{T&lt;:Integer, q}</code></pre><p>A mutable struct representing a sequence of integers with a maximum value constraint <code>q</code>.</p><ol><li><strong>Explicit Construction</strong>: <code>NumSequence(seq::AbstractVector{T}, q::Integer)</code> or <code>NumSequence{T,q}(seq)</code></li><li><strong>Random Construction</strong>: <code>NumSequence{T,q}(L::Integer)</code> or <code>NumSequence(L::Integer, q::Integer; T=IntType)</code>.  Construct a <code>NumSequence</code> of length <code>L</code> with random integers of type <code>T</code> in the range <code>[1, q]</code>.</li></ol><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; seq = [1, 2, 3, 4]
julia&gt; num_seq = NumSequence(seq, 4)

julia&gt; random_seq = NumSequence(10, 5; T=Int8)


julia&gt; max_value = num_seq.q  # Returns 4


julia&gt; copied_seq = copy(num_seq)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sequences.jl#L161-L184">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.PottsGraph"><a class="docstring-binding" href="#PottsEvolver.PottsGraph"><code>PottsEvolver.PottsGraph</code></a> — <span class="docstring-category">Type</span></summary><section><div><pre><code class="language-julia hljs">PottsGraph(L, q[, T]; init = :null, alphabet)</code></pre><p>Return a <code>PottsGraph{T}</code> of the required size.</p><ul><li><code>init == :null</code>: parameters are intialized to zero.</li><li><code>init == :rand</code>: parameters are randomly sampled using <code>Jrand</code> and <code>hrand</code> keywords.</li></ul><p><code>alphabet</code> is <code>aa_alphabet</code> if <code>q=21</code>, <code>nothing</code> otherwise.</p><p><strong>Random initialization</strong></p><ul><li><code>hrand</code> should be a function <code>q -&gt; h</code>.</li><li><code>Jrand</code> should be a function <code>q -&gt; J</code>.</li></ul><p><code>Jrand</code> does not have to return a symetric matrix. The output matrix is made symetric with zeroes on the diagonal blocks.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/pottsgraph.jl#L28-L45">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.PottsGraph"><a class="docstring-binding" href="#PottsEvolver.PottsGraph"><code>PottsEvolver.PottsGraph</code></a> — <span class="docstring-category">Type</span></summary><section><div><pre><code class="language-julia hljs">PottsGraph{T}</code></pre><ul><li>Array <code>J</code> of dimensions <code>q x q x L x L</code> and eltype <code>T</code></li><li>Array <code>h</code> of dimensions <code>q x L</code> and eltype <code>T</code></li><li>Inverse temperature <code>β</code></li><li><code>alphabet</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/pottsgraph.jl#L1-L8">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.SamplingParameters"><a class="docstring-binding" href="#PottsEvolver.SamplingParameters"><code>PottsEvolver.SamplingParameters</code></a> — <span class="docstring-category">Type</span></summary><section><div><pre><code class="language-julia hljs">mutable struct SamplingParameters</code></pre><p>Construct using keyword arguments:</p><pre><code class="language-julia hljs">sampling_type::Symbol = :discrete
step_type::Symbol = :gibbs
Teq::Real = 0 # time between samples, for chain
burnin::Real = 5*Teq # time before first sample
step_meaning::Symbol = :accepted # relevant for :discrete
fraction_gap_step::Float64 = 0.9 # relevant for :discrete and codon sampling
branchlength_meaning::BranchLengthMeaning # relevant for :discrete and sampling on a tree
substitution_rate::Union{Nothing,Float64} = nothing # relevant for :continuous
track_substitutions::Bool = false # relevant for continuous</code></pre><p><code>sampling_type</code> can be <code>:discrete</code> or <code>:continuous</code>. The time between samples is <code>Teq</code>, and an extra time <code>burnin</code> is taken before the first sample. Allowed values of <code>step_type</code> differ between the continuous and discrete cases.</p><p><strong>Discrete sampling</strong></p><ul><li><p><code>Teq</code> is measured in swaps: accepted (or attempted) changes of one sequence position. It must be an integer.</p></li><li><p><code>burnin</code>: number of steps starting from the initial sequence before the first sample. Must be an integer.</p></li><li><p><code>step_type</code>: <code>:gibbs</code> only. I intend to implement <code>:metropolis</code> and maybe <code>:glauber</code>.</p></li><li><p><code>step_meaning</code> (:discrete only): whether a swap count towards equilibration or not. It can take three values</p><ul><li><code>:proposed</code>: all mcmc steps count towards equilibration</li><li><code>:accepted</code>: only accepted steps count (all steps for non-codon Gibbs)</li><li><code>:changed</code>: only steps that lead to a change count (Gibbs can resample the same state)</li></ul><p><em>Note</em>: Gibbs steps for codons are more complicated, since they involve the possibility   Metropolis step for gaps, which can be rejected.</p></li><li><p><code>fraction_gap_step</code>: fraction of <code>:metropolis</code> steps concerning gaps when sampling with codons. This is only relevant for discrete sampling with codons.</p></li><li><p><code>branchlength_meaning</code>: how branch-lengths on a tree must be converted to swaps.   See <code>?BranchLengthMeaning</code> for information.   This is only useful if you sample along branches of a tree.</p></li></ul><p><strong>Continuous sampling</strong></p><p>For continuous sampling, there is no notion of swap/sweep or of accepted/rejected steps. Since any positive real number is acceptable as a sampling time, branch lenghts of trees can be used directly.</p><ul><li><code>Teq</code> and <code>burnin</code> are floats.</li><li><code>step_type</code> can be <code>:metropolis</code>, <code>:glauber</code> or <code>:sqrt</code>.</li><li><code>substitution_rate</code> is the average substitution rate for a given Potts model (the average runs over sequences). It is the result of <code>average_substitution_rate</code>. This is computed automatically if not provided (but takes some time).</li><li><code>track_substitutions</code>: track all substitutions (position, state, time) occuring during the Gillespie simulation. They are returned in the <code>info</code> output of <code>mcmc_sample</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sampling_parameters.jl#L72-L128">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.energy-Tuple{AbstractVector{&lt;:Integer}, PottsGraph}"><a class="docstring-binding" href="#PottsEvolver.energy-Tuple{AbstractVector{&lt;:Integer}, PottsGraph}"><code>PottsEvolver.energy</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">energy(s, g::PottsGraph)</code></pre><p>Energy of sequence <code>s</code> in <code>g</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/pottsgraph.jl#L174-L178">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.genetic_code-Tuple{PottsEvolver.Codon}"><a class="docstring-binding" href="#PottsEvolver.genetic_code-Tuple{PottsEvolver.Codon}"><code>PottsEvolver.genetic_code</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">genetic_code(c::Codon)</code></pre><p>Translate <code>c</code> and return the amino acid as a <code>Char</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/codons.jl#L99-L103">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.genetic_code-Tuple{T} where T&lt;:Integer"><a class="docstring-binding" href="#PottsEvolver.genetic_code-Tuple{T} where T&lt;:Integer"><code>PottsEvolver.genetic_code</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">genetic_code(x::Integer)</code></pre><p>Translate the <code>i</code>th codon and return the index of the corresponding amino acid, using the default <code>aa_alphabet</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/codons.jl#L89-L94">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.mcmc_sample"><a class="docstring-binding" href="#PottsEvolver.mcmc_sample"><code>PottsEvolver.mcmc_sample</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">mcmc_sample(
    g::PottsGraph, M::Integer, s0::AbstractSequence, params::SamplingParameters;
    rng=Random.GLOBAL_RNG, verbose=0, progress_meter=true, alignment_output=true,
)
mcmc_sample(
    g::PottsGraph, M::Integer, params::SamplingParameters; init=:random_num, kwargs...)
)
mcmc_sample(g, tvals::AbstractVector, s0, params::SamplingParameters; kwargs...)</code></pre><p>First form: sample <code>g</code> for <code>M</code> steps starting from <code>s0</code>, using parameters in <code>params</code>. Return value: named tuple with fields</p><ul><li><code>sequences</code>: alignment (or vector) of sequences</li><li><code>tvals</code>: vector with the number of steps at each sample</li><li><code>info</code>: information about the run</li><li><code>params</code>: parameters of the run.</li></ul><p>Second form: same, but initial sequence is provided through the <code>init</code> kwarg.</p><p>Third form: provide a set of times <code>tvals</code> at which samples are taken. Can also be used with the <code>init</code> kwarg.</p><p><em>Note</em>: this function is not very efficient if <code>M</code> is small.</p><p>Sampling details are determined by <code>parameters</code>, see <code>?SamplingParameters</code>. Whether to use the genetic code is determined by the type of the <code>init</code> sequence: it is used if <code>init::CodonSequence</code>, otherwise not.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sampling.jl#L5-L33">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.mcmc_sample-Tuple{Any, TreeTools.Tree, Any}"><a class="docstring-binding" href="#PottsEvolver.mcmc_sample-Tuple{Any, TreeTools.Tree, Any}"><code>PottsEvolver.mcmc_sample</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">mcmc_sample(g, tree, M=1, params; alignment_output, translate_output, init, kwargs...)</code></pre><p>Sample <code>g</code> along branches of <code>tree</code>. Repeat the process <code>M</code> times, returning an array of named tuples of the form   <code>(; tree, leaf_sequences, internal_sequences)</code>. If <code>M</code> is omitted, the output is just a named tuple (no array). Sequences in <code>leaf_sequences</code> and <code>internal_sequences</code> are sorted in post-order traversal.</p><p>The sequence to be used as the root should be provided using the <code>init</code> kwarg,   see <code>?PottsEvolver.get_init_sequence</code>.</p><p>If <code>alignment_output</code>, the sequences will be wrapped into an <code>Alignment</code> structure. Otherwise, they are in a dictionary indexed by node label. If <code>translate_output</code> and if the root sequence was a <code>CodonSequence</code>, the output alignment will contain the amino acid sequence and not the codons.</p><p><strong>Warning</strong></p><p>The <code>Teq</code> field of <code>params</code> is not used in the sampling. However, the <code>burnin</code> field will be used to set the root sequence: <code>burnin</code> mcmc steps will be performed starting from the input sequence, and the result is placed at the root. If you want a precise root sequence to be used, set <code>burnin=0</code> in <code>params</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/sampling.jl#L101-L123">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.read_graph"><a class="docstring-binding" href="#PottsEvolver.read_graph"><code>PottsEvolver.read_graph</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">read_graph
read_potts_graph</code></pre><p>Read <code>PottsGraph</code> object from a file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/IO.jl#L6-L11">source</a></section></details></article><article><details class="docstring" open="true"><summary id="PottsEvolver.read_potts_graph"><a class="docstring-binding" href="#PottsEvolver.read_potts_graph"><code>PottsEvolver.read_potts_graph</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">read_potts_graph</code></pre><p>Alias for <code>read_graph</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/PottsEvolver.jl/blob/c02e6cdd0267a03f753dfde8201b38cfead9ffcb/src/IO.jl#L13-L17">source</a></section></details></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../genetic_code/">« Genetic code</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Friday 9 January 2026 20:23">Friday 9 January 2026</span>. Using Julia version 1.12.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
